### PR TITLE
eth/tracers: Fix prestateTracer pre nonce on contract creation

### PR DESCRIPTION
**Pull request by andreisilviudragnea**
On contract creation, `prestateTracer` does not display the `pre` nonce for the contract account correctly. It should be `0`, not `1`, because it is incremented before the call to `CaptureStart` at https://github.com/ethereum/go-ethereum/blob/master/core/vm/evm.go#L450, so it needs to be decremented inside `prestateTracer` back to `0`.

Example call with `diffMode == true`:
```bash
curl --location 'http://0.0.0.0:8545' \
--header 'Content-Type: application/json' \
--data '{
    "jsonrpc": "2.0",
    "method": "debug_traceTransaction",
    "params": [
        "0x6908e9580db22a79ecc3f36bd0a6eaa496ca2e188dac64ac20b621d13e5aa59c",
        {
            "tracer": "prestateTracer",
            "tracerConfig": {
                "diffMode": true
            }
        }
    ],
    "id": 1234
}'
```

Result before bug fix:
```json
{
    "jsonrpc": "2.0",
    "id": 1234,
    "result": {
        "post": {
            "0x1bda2f8e4735507930bd6cfe873bf0bf0f4ab1de": {
                "code": "0x608060405234801561001057600080fd5b50600436106100365760003560e01c806309ce9ccb1461003b5780633fb5c1cb14610059575b600080fd5b610043610075565b60405161005091906100e2565b60405180910390f35b610073600480360381019061006e919061012e565b61007b565b005b60005481565b80600081905550600a8111156100c6576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016100bd906101de565b60405180910390fd5b50565b6000819050919050565b6100dc816100c9565b82525050565b60006020820190506100f760008301846100d3565b92915050565b600080fd5b61010b816100c9565b811461011657600080fd5b50565b60008135905061012881610102565b92915050565b600060208284031215610144576101436100fd565b5b600061015284828501610119565b91505092915050565b600082825260208201905092915050565b7f4e756d6265722069732067726561746572207468616e2031302c207472616e7360008201527f616374696f6e2072657665727465642e00000000000000000000000000000000602082015250565b60006101c860308361015b565b91506101d38261016c565b604082019050919050565b600060208201905081810360008301526101f7816101bb565b905091905056fea264697066735822122069018995fecf03bda91a88b6eafe41641709dee8b4a706fe301c8a569fe8c1b364736f6c63430008130033"
            },
            "0x2445e8c26a2bf3d1e59f1bb9b1d442caf90768e0": {
                "balance": "0x10f0645688331eb5690"
            },
            "0x82211934c340b29561381392348d48413e15adc8": {
                "balance": "0x6aae9b21b6ee855",
                "nonce": 65
            }
        },
        "pre": {
            "0x1bda2f8e4735507930bd6cfe873bf0bf0f4ab1de": {
                "balance": "0x0",
                "nonce": 1
            },
            "0x2445e8c26a2bf3d1e59f1bb9b1d442caf90768e0": {
                "balance": "0x10f06447a8d44dba190",
                "nonce": 2
            },
            "0x82211934c340b29561381392348d48413e15adc8": {
                "balance": "0x6abd7a808913ed2",
                "nonce": 64
            }
        }
    }
}
```

Result after bug fix:
```json
{
    "jsonrpc": "2.0",
    "id": 1234,
    "result": {
        "post": {
            "0x1bda2f8e4735507930bd6cfe873bf0bf0f4ab1de": {
                "code": "0x608060405234801561001057600080fd5b50600436106100365760003560e01c806309ce9ccb1461003b5780633fb5c1cb14610059575b600080fd5b610043610075565b60405161005091906100e2565b60405180910390f35b610073600480360381019061006e919061012e565b61007b565b005b60005481565b80600081905550600a8111156100c6576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016100bd906101de565b60405180910390fd5b50565b6000819050919050565b6100dc816100c9565b82525050565b60006020820190506100f760008301846100d3565b92915050565b600080fd5b61010b816100c9565b811461011657600080fd5b50565b60008135905061012881610102565b92915050565b600060208284031215610144576101436100fd565b5b600061015284828501610119565b91505092915050565b600082825260208201905092915050565b7f4e756d6265722069732067726561746572207468616e2031302c207472616e7360008201527f616374696f6e2072657665727465642e00000000000000000000000000000000602082015250565b60006101c860308361015b565b91506101d38261016c565b604082019050919050565b600060208201905081810360008301526101f7816101bb565b905091905056fea264697066735822122069018995fecf03bda91a88b6eafe41641709dee8b4a706fe301c8a569fe8c1b364736f6c63430008130033",
                "nonce": 1
            },
            "0x2445e8c26a2bf3d1e59f1bb9b1d442caf90768e0": {
                "balance": "0x10f0645688331eb5690"
            },
            "0x82211934c340b29561381392348d48413e15adc8": {
                "balance": "0x6aae9b21b6ee855",
                "nonce": 65
            }
        },
        "pre": {
            "0x2445e8c26a2bf3d1e59f1bb9b1d442caf90768e0": {
                "balance": "0x10f06447a8d44dba190",
                "nonce": 2
            },
            "0x82211934c340b29561381392348d48413e15adc8": {
                "balance": "0x6abd7a808913ed2",
                "nonce": 64
            }
        }
    }
}
```

Example call without `diffMode == true`:
```bash
curl --location 'http://0.0.0.0:8545' \
--header 'Content-Type: application/json' \
--data '{
    "jsonrpc": "2.0",
    "method": "debug_traceTransaction",
    "params": [
        "0x6908e9580db22a79ecc3f36bd0a6eaa496ca2e188dac64ac20b621d13e5aa59c",
        {
            "tracer": "prestateTracer"
        }
    ],
    "id": 1234
}'
```
Result before bug fix:
```json
{
    "jsonrpc": "2.0",
    "id": 1234,
    "result": {
        "0x1bda2f8e4735507930bd6cfe873bf0bf0f4ab1de": {
            "balance": "0x0",
            "nonce": 1
        },
        "0x2445e8c26a2bf3d1e59f1bb9b1d442caf90768e0": {
            "balance": "0x10f06447a8d44dba190",
            "nonce": 2
        },
        "0x82211934c340b29561381392348d48413e15adc8": {
            "balance": "0x6abd7a808913ed2",
            "nonce": 64
        }
    }
}
```
Result after bug fix:
```json
{
    "jsonrpc": "2.0",
    "id": 1234,
    "result": {
        "0x2445e8c26a2bf3d1e59f1bb9b1d442caf90768e0": {
            "balance": "0x10f06447a8d44dba190",
            "nonce": 2
        },
        "0x82211934c340b29561381392348d48413e15adc8": {
            "balance": "0x6abd7a808913ed2",
            "nonce": 64
        }
    }
}
```

### CODE REVIEW

**Review by s1na — CHANGES_REQUESTED** [review]
This will be fixed as part of #27629. But I'm for fixing it on master before. Only can you please add a test case for it? You can load this file into the console and pass it a tx that contains this condition: https://github.com/ethereum/go-ethereum/blob/490efc5f21ebde5d54a630fc7ff56cbd686c1f74/eth/tracers/internal/tracetest/makeTest.js

**Comment by andreisilviudragnea** [comment]
@s1na I added the required tests, but needed to make `ChainRules` public, because this change is dependent on `EIP-158`. Existing tests started failing because of my change being always applied.

File: eth/tracers/native/prestate.go [hunk 1]
```diff
@@ -102,6 +102,9 @@ func (t *prestateTracer) CaptureStart(env *vm.EVM, from common.Address, to commo
 	// The recipient balance includes the value transferred.
 	toBal := new(big.Int).Sub(t.pre[to].Balance, value)
 	t.pre[to].Balance = toBal
+	if env.ChainRules.IsEIP158 && create {
```
Comment (s1na) [inline eth/tracers/native/prestate.go:105]: Noo you don't need to expose ChainRules there is another way of getting it:

```go
	rules := env.ChainConfig().Rules(env.Context.BlockNumber, env.Context.Random != nil, env.Context.Time)
	```

Comment (andreisilviudragnea) [reply eth/tracers/native/prestate.go]: Done.

**Review by s1na — APPROVED** [review]
LGTM!

**Pull request merged**